File: utils/filesystem_manipulation/split_files_into_folders_by_created_date.py
# soledad penades www.soledadpenades.com

import os
import shutil
import sys
import time
from datetime import datetime

if len(sys.argv) > 1:
	folder = sys.argv[1]
else:
	folder = '.'

for item in os.listdir(folder):

	full_path = os.path.join(folder, item)
	
	if os.path.isdir(full_path):
		continue
	
        try:
            ctime = os.path.getctime(full_path)
        except OSError:
            ctime = 0

        try:
            mtime = os.path.getmtime(full_path)
        except OSError:
            mtime = 0

        final_time = min(ctime, mtime)

        created_date = time.ctime(ctime)
        modified_date = time.ctime(mtime)
        fu = time.ctime(final_time)
        d = datetime.strptime(fu,"%a %b %d %H:%M:%S %Y")
        final_date = d.strftime('%Y%m%d')
        
        print (created_date, modified_date, final_date)
        subfolder = final_date

	dst_folder = os.path.join(folder, subfolder)
	
	if not os.path.exists(dst_folder):
		os.mkdir(dst_folder)
	
        new_path = os.path.join(dst_folder, item)
        print (full_path, "=>", new_path)
        
	shutil.move(full_path, os.path.join(dst_folder, item))

File: utils/filesystem_manipulation/split_files_into_folders_by_created_date.js
// soledad penades www.soledadpenades.com

const process = require('node:process');
const path = require('node:path');
const fs = require('node:fs');
const argv = process.argv;

let folder;

// something something I don't need like .../bin/node
if(process.execPath === argv[0]) {
	argv.splice(0, 1);
}

// remaining is this script name and hopefully a destination folder name
if(argv.length > 1) {
	folder = argv[1];
} else {
	console.error("Specify destination folder to be split");
	process.exit(-1);
}

console.log(`Splitting ${folder}`);

let dstPath = folder;

if(!path.isAbsolute(dstPath)) {
	dstPath = path.resolve('.', dstPath);
	console.log(`Resolving to ${dstPath}`);
}

fs.readdirSync(dstPath, { withFileTypes: true }).forEach((item) => {
	let itemPath = path.join(dstPath, item.name);
	console.log(itemPath);

	// Leave directories alone - move only files
	if(item.isDirectory()) {
		console.log(`Skipping ${item.name} as it's a directory`);
		return;
	} 
	
	let s = fs.statSync(itemPath);
	let ctime = s.ctimeMs;
	let mtime = s.mtimeMs;
	let finalTime;

	if(ctime > mtime) {
		finalTime = s.mtime;
	} else {
		finalTime = s.ctime;
	}

	//let finalTime = Math.min(s.ctimeMs, s.mtimeMs);
	let finalDate = new Date(finalTime);
	let day = finalDate.getDate();
	let month = finalDate.getMonth() + 1; // 0 index based
	let year = finalDate.getFullYear();

	day = String(day).padStart(2, '0');
	month = String(month).padStart(2, '0');

	let subFolder = `${year}${month}${day}`;
	
	console.log(finalDate, year, month, day, subFolder);

	let subFolderPath = path.join(dstPath, subFolder);

	if(!fs.existsSync(subFolderPath)) {
		console.log(`${subFolder} does not exist yet, creating`);
		fs.mkdirSync(subFolderPath);
	}

	let newItemPath = path.join(subFolderPath, item.name);

	console.log(`${itemPath} => ${newItemPath}`);

	fs.renameSync(itemPath, newItemPath);

});
